File: sentil-java/src/main/java/io/github/sedislab/sentil/NoiseModel.java
package io.github.sedislab.sentil;

import java.util.OptionalDouble;

/** A noise distribution for stochastic signal lifting. */
public final class NoiseModel extends NativeResource {
    NoiseModel(long handle) {
        super(handle, NativeLib::noiseDestroy);
    }

    /** A point mass at value. */
    public static NoiseModel dirac(double value) throws SentilException {
        return new NoiseModel(NativeLib.noiseDirac(value));
    }

    /** A normal distribution. */
    public static NoiseModel gaussian(double mean, double stdDev) throws SentilException {
        return new NoiseModel(NativeLib.noiseGaussian(mean, stdDev));
    }

    /** A uniform distribution on [low, high]. */
    public static NoiseModel uniform(double low, double high) throws SentilException {
        return new NoiseModel(NativeLib.noiseUniform(low, high));
    }

    /** A log-normal distribution. */
    public static NoiseModel logNormal(double mu, double sigma) throws SentilException {
        return new NoiseModel(NativeLib.noiseLogNormal(mu, sigma));
    }

    /** An exponential distribution with the given rate. */
    public static NoiseModel exponential(double rate) throws SentilException {
        return new NoiseModel(NativeLib.noiseExponential(rate));
    }

    /** A gamma distribution. */
    public static NoiseModel gamma(double shape, double scale) throws SentilException {
        return new NoiseModel(NativeLib.noiseGamma(shape, scale));
    }

    /** A beta distribution. */
    public static NoiseModel beta(double alpha, double beta) throws SentilException {
        return new NoiseModel(NativeLib.noiseBeta(alpha, beta));
    }

    /** A Weibull distribution. */
    public static NoiseModel weibull(double shape, double scale) throws SentilException {
        return new NoiseModel(NativeLib.noiseWeibull(shape, scale));
    }

    /** A Rayleigh distribution. */
    public static NoiseModel rayleigh(double scale) throws SentilException {
        return new NoiseModel(NativeLib.noiseRayleigh(scale));
    }

    /** A Gumbel distribution. */
    public static NoiseModel gumbel(double location, double scale) throws SentilException {
        return new NoiseModel(NativeLib.noiseGumbel(location, scale));
    }

    /** A Cauchy distribution. */
    public static NoiseModel cauchy(double location, double scale) throws SentilException {
        return new NoiseModel(NativeLib.noiseCauchy(location, scale));
    }

    /** A Student's t distribution. */
    public static NoiseModel studentT(double df, double location, double scale)
            throws SentilException {
        return new NoiseModel(NativeLib.noiseStudentT(df, location, scale));
    }

    /** A normal distribution truncated to [lower, upper]. */
    public static NoiseModel truncatedNormal(double mean, double stdDev, double lower, double upper)
            throws SentilException {
        return new NoiseModel(NativeLib.noiseTruncatedNormal(mean, stdDev, lower, upper));
    }

    /** A Poisson distribution with the given rate. */
    public static NoiseModel poisson(double rate) throws SentilException {
        return new NoiseModel(NativeLib.noisePoisson(rate));
    }

    /** A binomial distribution of n trials with success probability p. */
    public static NoiseModel binomial(long n, double p) throws SentilException {
        return new NoiseModel(NativeLib.noiseBinomial(n, p));
    }

    /** An empirical model resampled from residuals. */
    public static NoiseModel bootstrap(double[] residuals) throws SentilException {
        return new NoiseModel(NativeLib.noiseBootstrap(residuals));
    }

    /** Rebuild a model from JSON. */
    public static NoiseModel fromJson(String json) throws SentilException {
        return new NoiseModel(NativeLib.noiseFromJson(json));
    }

    /** Load a model from a JSON file. */
    public static NoiseModel fromFile(String path) throws SentilException {
        return new NoiseModel(NativeLib.noiseFromFile(path));
    }

    /** The analytic mean, or empty where it is undefined. */
    public OptionalDouble mean() {
        double[] result = NativeLib.noiseMean(handle());
        return result.length == 0 ? OptionalDouble.empty() : OptionalDouble.of(result[0]);
    }

    /** The analytic variance, or empty where it is undefined. */
    public OptionalDouble variance() {
        double[] result = NativeLib.noiseVariance(handle());
        return result.length == 0 ? OptionalDouble.empty() : OptionalDouble.of(result[0]);
    }

    /** The model as a JSON string. */
    public String toJson() throws SentilException {
        return NativeLib.noiseToJson(handle());
    }
}

File: sentil-matlab/+sentil/NoiseModel.m
classdef NoiseModel < handle
    %NOISEMODEL A sensor noise distribution.

    properties (SetAccess = private, Hidden)
        Handle uint64 = uint64(0)
    end

    methods (Static)
        function n = dirac(value)
            n = sentil.NoiseModel(sentil_mex('noise_dirac', double(value)));
        end
        function n = gaussian(mean, stdDev)
            n = sentil.NoiseModel(sentil_mex('noise_gaussian', double(mean), double(stdDev)));
        end
        function n = uniform(low, high)
            n = sentil.NoiseModel(sentil_mex('noise_uniform', double(low), double(high)));
        end
        function n = log_normal(mu, sigma)
            n = sentil.NoiseModel(sentil_mex('noise_log_normal', double(mu), double(sigma)));
        end
        function n = exponential(lambda)
            n = sentil.NoiseModel(sentil_mex('noise_exponential', double(lambda)));
        end
        function n = gamma(shape, scale)
            n = sentil.NoiseModel(sentil_mex('noise_gamma', double(shape), double(scale)));
        end
        function n = beta(alpha, betaParam)
            n = sentil.NoiseModel(sentil_mex('noise_beta', double(alpha), double(betaParam)));
        end
        function n = weibull(shape, scale)
            n = sentil.NoiseModel(sentil_mex('noise_weibull', double(shape), double(scale)));
        end
        function n = rayleigh(scale)
            n = sentil.NoiseModel(sentil_mex('noise_rayleigh', double(scale)));
        end
        function n = gumbel(location, scale)
            n = sentil.NoiseModel(sentil_mex('noise_gumbel', double(location), double(scale)));
        end
        function n = cauchy(location, scale)
            n = sentil.NoiseModel(sentil_mex('noise_cauchy', double(location), double(scale)));
        end
        function n = student_t(df, location, scale)
            n = sentil.NoiseModel(sentil_mex('noise_student_t', double(df), double(location), ...
                double(scale)));
        end
        function n = truncated_normal(mean, stdDev, lower, upper)
            n = sentil.NoiseModel(sentil_mex('noise_truncated_normal', double(mean), ...
                double(stdDev), double(lower), double(upper)));
        end
        function n = poisson(lambda)
            n = sentil.NoiseModel(sentil_mex('noise_poisson', double(lambda)));
        end
        function n = binomial(trials, p)
            n = sentil.NoiseModel(sentil_mex('noise_binomial', double(trials), double(p)));
        end
        function n = bootstrap(residuals)
            %BOOTSTRAP An empirical model resampled from residuals.
            n = sentil.NoiseModel(sentil_mex('noise_bootstrap', double(residuals)));
        end
        function n = mixture(weights, models)
            %MIXTURE A weighted mixture. The component models are consumed.
            handles = zeros(1, numel(models), 'uint64');
            for i = 1:numel(models)
                handles(i) = models(i).consume();
            end
            n = sentil.NoiseModel(sentil_mex('noise_mixture', double(weights), handles));
        end
        function n = fit_gaussian(samples)
            n = sentil.NoiseModel(sentil_mex('noise_fit_gaussian', double(samples)));
        end
        function n = fit_bootstrap(samples)
            n = sentil.NoiseModel(sentil_mex('noise_fit_bootstrap', double(samples)));
        end
        function n = fit_bootstrap_reservoir(samples, maxSamples)
            n = sentil.NoiseModel(sentil_mex('noise_fit_bootstrap_reservoir', double(samples), ...
                double(maxSamples)));
        end
        function n = fit_gaussian_mixture(samples, components, maxIters)
            n = sentil.NoiseModel(sentil_mex('noise_fit_gaussian_mixture', double(samples), ...
                double(components), double(maxIters)));
        end
        function n = from_json(json)
            n = sentil.NoiseModel(sentil_mex('noise_from_json', json));
        end
        function n = from_file(path)
            n = sentil.NoiseModel(sentil_mex('noise_from_file', path));
        end

        function r = residuals(groundTruth, sensor, interaction)
            %RESIDUALS The residuals between paired truth and sensor readings.
            if nargin < 3
                interaction = sentil.NoiseInteraction.Additive;
            end
            r = sentil_mex('noise_residuals', double(groundTruth), double(sensor), ...
                double(int32(interaction)));
        end
    end

    methods
        function obj = NoiseModel(handle)
            obj.Handle = handle;
        end

        function delete(obj)
            if obj.Handle ~= 0
                sentil_mex('noise_destroy', obj.Handle);
                obj.Handle = uint64(0);
            end
        end

        function m = mean(obj)
            %MEAN The analytic mean, or [] when undefined.
            obj.assertOpen();
            m = sentil_mex('noise_mean', obj.Handle);
        end

        function v = variance(obj)
            %VARIANCE The analytic variance, or [] when undefined.
            obj.assertOpen();
            v = sentil_mex('noise_variance', obj.Handle);
        end

        function s = to_json(obj)
            %TO_JSON The model as a JSON string.
            obj.assertOpen();
            s = sentil_mex('noise_to_json', obj.Handle);
        end
    end

    methods (Hidden)
        function h = consume(obj)
            %CONSUME Surrender the native handle, leaving this model closed.
            obj.assertOpen();
            h = obj.Handle;
            obj.Handle = uint64(0);
        end
    end

    methods (Access = private)
        function assertOpen(obj)
            if obj.Handle == 0
                error('sentil:handle', 'this noise model has been closed');
            end
        end
    end
end